File: ch15/ch15_src/src/bin/shm_ts.rs
use std::{process::exit, ptr::null_mut};

use nix::{libc::{self, c_char, ctime_r, shmat, shmctl, shmget, IPC_CREAT, IPC_RMID}, unistd::sleep};

const TIME_MEM_KEY:i32 = 99;
const SEG_SIZE:usize = 100;

fn main(){
    let seg_id = unsafe { shmget(TIME_MEM_KEY, SEG_SIZE, IPC_CREAT|0o777) };
    if seg_id==-1 {
        eprintln!("shmget");
        exit(1);
    }

    let mem_ptr =  unsafe { shmat(seg_id, std::ptr::null(), 0)  as *mut c_char};
    if unsafe { *mem_ptr } == -1  {
      eprintln!("shmat");
      exit(2);
    }

    for _ in 0..60{
      let now:i64 = 0;
      unsafe { libc::time(now as *mut i64) };
      unsafe { ctime_r(now as *const i64, mem_ptr) };
      sleep(1);
    }

    unsafe { shmctl(seg_id, IPC_RMID, null_mut()) };
}

File: ch15/shm_ts.c
#include <stdio.h>
#include <string.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <time.h>
#include <unistd.h>
#include <stdlib.h>

#define TIME_MEM_KEY 99
#define SEG_SIZE ((size_t) 100)
#define oops(m, x)                                                             \
  {                                                                            \
    perror(m);                                                                 \
    exit(x);                                                                   \
  }

int main(){
  int seg_id;
  char *mem_ptr;
  long now;
  int n;

  seg_id = shmget(TIME_MEM_KEY, SEG_SIZE, IPC_CREAT|0777);
  if (seg_id==-1) {
    oops("shmget", 1);
  }

  mem_ptr = shmat(seg_id, NULL, 0);
  if (mem_ptr==(void*) -1) {
    oops("shmat", 2);
  }

  for (n=0; n<60;n++) {
    time(&now);
    strcpy(mem_ptr, ctime(&now));
    sleep(1);
  }
  shmctl(seg_id, IPC_RMID, NULL);
}